File: src/cccc/ports/mcp/utils/help_markdown.py
from __future__ import annotations

import re
from typing import Any, Optional


_HELP_ROLE_HEADER_RE = re.compile(r"^##\s*@role:\s*(\w+)\s*$", re.IGNORECASE)
_HELP_ACTOR_HEADER_RE = re.compile(r"^##\s*@actor:\s*(\S+)(?:\s+(.*\S))?\s*$", re.IGNORECASE)
_HELP_PET_HEADER_RE = re.compile(r"^##\s*@pet\s*:?\s*$", re.IGNORECASE)
_HELP_VOICE_SECRETARY_HEADER_RE = re.compile(r"^##\s*@voice_secretary\s*:?\s*$", re.IGNORECASE)
_HELP_H2_RE = re.compile(r"^##(?!#)\s+.*$")


def _split_sections(markdown: str) -> list[str]:
    raw = str(markdown or "").replace("\r\n", "\n").replace("\r", "\n")
    if not raw:
        return [""]
    lines = raw.split("\n")
    sections: list[str] = []
    current: list[str] = []
    for line in lines:
        if _HELP_H2_RE.match(line) and current:
            sections.append("\n".join(current))
            current = [line]
            continue
        current.append(line)
    sections.append("\n".join(current))
    return sections


def _trim_block(text: str) -> str:
    return str(text or "").strip()


def _parse_tagged_section(section: str) -> Optional[dict[str, str]]:
    normalized = str(section or "").replace("\r\n", "\n").replace("\r", "\n")
    lines = normalized.split("\n")
    header = str(lines[0] or "")
    role_match = _HELP_ROLE_HEADER_RE.match(header)
    if role_match:
        role = str(role_match.group(1) or "").strip().lower()
        body = _trim_block("\n".join(lines[1:]))
        if role in {"foreman", "peer"}:
            return {"kind": "role", "key": f"role:{role}", "raw": _trim_block(normalized), "body": body}
        return {"kind": "extra", "key": f"role:{role}", "raw": _trim_block(normalized), "body": body}
    actor_match = _HELP_ACTOR_HEADER_RE.match(header)
    if actor_match:
        actor_id = str(actor_match.group(1) or "").strip()
        inline_body = str(actor_match.group(2) or "").strip()
        body_lines = [inline_body] if inline_body else []
        body_lines.extend(lines[1:])
        return {
            "kind": "actor" if actor_id else "extra",
            "key": f"actor:{actor_id}" if actor_id else "actor:",
            "raw": _trim_block(normalized),
            "body": _trim_block("\n".join(body_lines)),
        }
    if _HELP_PET_HEADER_RE.match(header):
        return {
            "kind": "extra",
            "key": "legacy:pet",
            "raw": _trim_block(normalized),
            "body": _trim_block("\n".join(lines[1:])),
        }
    if _HELP_VOICE_SECRETARY_HEADER_RE.match(header):
        return {
            "kind": "voice_secretary",
            "key": "voice_secretary",
            "raw": _trim_block(normalized),
            "body": _trim_block("\n".join(lines[1:])),
        }
    return None


def parse_help_markdown(markdown: str) -> dict[str, Any]:
    sections = _split_sections(markdown)
    common_sections: list[str] = []
    actor_notes: dict[str, str] = {}
    extra_tagged_blocks: list[str] = []
    foreman = ""
    peer = ""
    voice_secretary = ""

    for section in sections:
        raw = _trim_block(section)
        if not raw:
            continue
        tagged = _parse_tagged_section(raw)
        if not tagged:
            common_sections.append(raw)
            continue
        kind = str(tagged.get("kind") or "")
        key = str(tagged.get("key") or "")
        body = str(tagged.get("body") or "")
        if kind == "role":
            if key == "role:foreman":
                foreman = body
            elif key == "role:peer":
                peer = body
            else:
                extra_tagged_blocks.append(str(tagged.get("raw") or ""))
            continue
        if kind == "actor":
            actor_id = key[len("actor:") :]
            if actor_id:
                actor_notes[actor_id] = body
            else:
                extra_tagged_blocks.append(str(tagged.get("raw") or ""))
            continue
        if kind == "voice_secretary":
            voice_secretary = body
            continue
        extra_tagged_blocks.append(str(tagged.get("raw") or ""))

    return {
        "common": "\n\n".join(common_sections).strip(),
        "foreman": foreman,
        "peer": peer,
        "voice_secretary": voice_secretary,
        "actor_notes": actor_notes,
        "extra_tagged_blocks": extra_tagged_blocks,
    }


def build_help_markdown(
    *,
    common: str,
    foreman: str,
    peer: str,
    actor_notes: dict[str, str],
    voice_secretary: str = "",
    actor_order: Optional[list[str]] = None,
    extra_tagged_blocks: Optional[list[str]] = None,
) -> str:
    parts: list[str] = []
    common_text = _trim_block(common)
    foreman_text = _trim_block(foreman)
    peer_text = _trim_block(peer)
    voice_secretary_text = _trim_block(voice_secretary)
    actor_notes_map = dict(actor_notes or {})
    extra_blocks = [_trim_block(item) for item in list(extra_tagged_blocks or []) if _trim_block(item)]

    if common_text:
        parts.append(common_text)
    if foreman_text:
        parts.append(f"## @role: foreman\n\n{foreman_text}")
    if peer_text:
        parts.append(f"## @role: peer\n\n{peer_text}")
    if voice_secretary_text:
        parts.append(f"## @voice_secretary\n\n{voice_secretary_text}")

    seen: set[str] = set()
    ordered_actor_ids: list[str] = []
    for actor_id in list(actor_order or []):
        aid = str(actor_id or "").strip()
        if not aid or aid in seen:
            continue
        seen.add(aid)
        ordered_actor_ids.append(aid)
    for actor_id in sorted(actor_notes_map.keys()):
        if actor_id in seen:
            continue
        seen.add(actor_id)
        ordered_actor_ids.append(actor_id)
    for actor_id in ordered_actor_ids:
        body = _trim_block(str(actor_notes_map.get(actor_id) or ""))
        if not body:
            continue
        parts.append(f"## @actor: {actor_id}\n\n{body}")

    parts.extend(extra_blocks)
    out = "\n\n".join([part for part in parts if part]).strip()
    return f"{out}\n" if out else ""


def update_actor_help_note(markdown: str, actor_id: str, note: str, actor_order: Optional[list[str]] = None) -> str:
    parsed = parse_help_markdown(markdown)
    next_actor_notes = dict(parsed.get("actor_notes") or {})
    aid = str(actor_id or "").strip()
    if aid:
        next_actor_notes[aid] = _trim_block(note)
    if aid and not next_actor_notes.get(aid):
        next_actor_notes.pop(aid, None)
    return build_help_markdown(
        common=str(parsed.get("common") or ""),
        foreman=str(parsed.get("foreman") or ""),
        peer=str(parsed.get("peer") or ""),
        voice_secretary=str(parsed.get("voice_secretary") or ""),
        actor_notes=next_actor_notes,
        actor_order=actor_order,
        extra_tagged_blocks=list(parsed.get("extra_tagged_blocks") or []),
    )


def _select_help_markdown(
    markdown: str,
    *,
    role: Optional[str],
    actor_id: Optional[str],
    include_voice_secretary: bool = False,
) -> str:
    """Filter CCCC_HELP markdown by optional conditional blocks.

    Supported markers (level-2 headings):
    - "## @role: foreman|peer"
    - "## @actor: <actor_id>"
    - "## @voice_secretary"

    Untagged content is always included. Tagged blocks are filtered only when the selector is known.

    A tagged block starts at its marker heading and ends at the next level-2 heading.
    Within tagged blocks, prefer "###" for subheadings (so "##" can remain a block boundary).
    """
    raw = str(markdown or "")
    if not raw.strip():
        return raw

    role_norm = str(role or "").strip().casefold()
    actor_norm = str(actor_id or "").strip()
    lines = raw.splitlines()
    keep_trailing_newline = raw.endswith("\n")

    out: list[str] = []
    buf: list[str] = []
    tag_kind: Optional[str] = None
    tag_value: str = ""

    def _include_block() -> bool:
        if tag_kind is None:
            return True
        if tag_kind == "role":
            if not role_norm:
                return True
            return role_norm == str(tag_value or "").strip().casefold()
        if tag_kind == "actor":
            if not actor_norm:
                return False
            return actor_norm == str(tag_value or "").strip()
        if tag_kind == "legacy_pet":
            return False
        if tag_kind == "voice_secretary":
            return bool(include_voice_secretary)
        return True

    def _flush() -> None:
        nonlocal buf
        if buf and _include_block():
            out.extend(buf)
        buf = []

    for ln in lines:
        m_role = _HELP_ROLE_HEADER_RE.match(ln)
        m_actor = _HELP_ACTOR_HEADER_RE.match(ln)
        m_pet = _HELP_PET_HEADER_RE.match(ln)
        m_voice_secretary = _HELP_VOICE_SECRETARY_HEADER_RE.match(ln)
        is_h2 = bool(_HELP_H2_RE.match(ln))

        if m_role or m_actor or m_pet or m_voice_secretary:
            _flush()
            if m_role:
                tag_kind = "role"
                tag_value = str(m_role.group(1) or "").strip()
                role_label = tag_value.strip().casefold()
                if role_label == "foreman":
                    ln = "## Foreman"
                elif role_label == "peer":
                    ln = "## Peer"
                else:
                    ln = f"## Role: {tag_value}"
            else:
                if m_actor:
                    tag_kind = "actor"
                    tag_value = str(m_actor.group(1) or "").strip()
                    ln = "## Notes for you"
                elif m_pet:
                    tag_kind = "legacy_pet"
                    tag_value = ""
                else:
                    tag_kind = "voice_secretary"
                    tag_value = ""
                    ln = "## Voice Secretary Operating Contract"
            buf.append(ln)
            continue

        if is_h2:
            _flush()
            tag_kind = None
            tag_value = ""

        buf.append(ln)
    _flush()

    out_text = "\n".join(out)
    if keep_trailing_newline:
        out_text += "\n"
    return out_text

File: web/src/utils/helpMarkdown.ts
export type HelpChangedBlock = "common" | "role:foreman" | "role:peer" | "voice_secretary" | `actor:${string}`;

export type ParsedHelpMarkdown = {
  common: string;
  foreman: string;
  peer: string;
  voiceSecretary: string;
  actorNotes: Record<string, string>;
  extraTaggedBlocks: string[];
};

type TaggedSection = {
  kind: "role" | "actor" | "voice_secretary" | "extra";
  key: string;
  raw: string;
  body: string;
};

const H2_RE = /^##(?!#)\s+.*$/;
const ROLE_TAG_RE = /^##\s*@role:\s*(\w+)\s*$/i;
const ACTOR_TAG_RE = /^##\s*@actor:\s*(\S+)(?:\s+(.*\S))?\s*$/i;
const PET_TAG_RE = /^##\s*@pet\s*:?\s*$/i;
const VOICE_SECRETARY_TAG_RE = /^##\s*@voice_secretary\s*:?\s*$/i;

function splitSections(markdown: string): string[] {
  const raw = String(markdown || "").replace(/\r\n?/g, "\n");
  if (!raw) return [""];
  const lines = raw.split("\n");
  const sections: string[] = [];
  let current: string[] = [];
  for (const line of lines) {
    if (H2_RE.test(line) && current.length > 0) {
      sections.push(current.join("\n"));
      current = [line];
      continue;
    }
    current.push(line);
  }
  sections.push(current.join("\n"));
  return sections;
}

function trimBlock(text: string): string {
  return String(text || "").trim();
}

function parseTaggedSection(section: string): TaggedSection | null {
  const normalized = String(section || "").replace(/\r\n?/g, "\n");
  const lines = normalized.split("\n");
  const header = String(lines[0] || "");
  const roleMatch = header.match(ROLE_TAG_RE);
  if (roleMatch) {
    const role = String(roleMatch[1] || "").trim().toLowerCase();
    const body = trimBlock(lines.slice(1).join("\n"));
    if (role === "foreman" || role === "peer") {
      return { kind: "role", key: `role:${role}`, raw: trimBlock(normalized), body };
    }
    return { kind: "extra", key: `role:${role}`, raw: trimBlock(normalized), body };
  }
  const actorMatch = header.match(ACTOR_TAG_RE);
  if (actorMatch) {
    const actorId = String(actorMatch[1] || "").trim();
    const inlineBody = String(actorMatch[2] || "").trim();
    const bodyLines = lines.slice(1);
    const body = trimBlock([inlineBody, ...bodyLines].filter(Boolean).join("\n"));
    return {
      kind: actorId ? "actor" : "extra",
      key: actorId ? `actor:${actorId}` : "actor:",
      raw: trimBlock(normalized),
      body,
    };
  }
  if (PET_TAG_RE.test(header)) {
    return {
      kind: "extra",
      key: "legacy:pet",
      raw: trimBlock(normalized),
      body: trimBlock(lines.slice(1).join("\n")),
    };
  }
  if (VOICE_SECRETARY_TAG_RE.test(header)) {
    return {
      kind: "voice_secretary",
      key: "voice_secretary",
      raw: trimBlock(normalized),
      body: trimBlock(lines.slice(1).join("\n")),
    };
  }
  return null;
}

export function parseHelpMarkdown(markdown: string): ParsedHelpMarkdown {
  const sections = splitSections(markdown);
  const commonSections: string[] = [];
  const actorNotes: Record<string, string> = {};
  const extraTaggedBlocks: string[] = [];
  let foreman = "";
  let peer = "";
  let voiceSecretary = "";

  for (const section of sections) {
    const raw = trimBlock(section);
    if (!raw) continue;
    const tagged = parseTaggedSection(raw);
    if (!tagged) {
      commonSections.push(raw);
      continue;
    }
    if (tagged.kind === "role") {
      if (tagged.key === "role:foreman") foreman = tagged.body;
      else if (tagged.key === "role:peer") peer = tagged.body;
      else extraTaggedBlocks.push(tagged.raw);
      continue;
    }
    if (tagged.kind === "actor") {
      const actorId = tagged.key.slice("actor:".length);
      if (actorId) actorNotes[actorId] = tagged.body;
      else extraTaggedBlocks.push(tagged.raw);
      continue;
    }
    if (tagged.kind === "voice_secretary") {
      voiceSecretary = tagged.body;
      continue;
    }
    extraTaggedBlocks.push(tagged.raw);
  }

  const common = commonSections.join("\n\n").trim();
  return { common, foreman, peer, voiceSecretary, actorNotes, extraTaggedBlocks };
}

export function buildHelpMarkdown(input: {
  common: string;
  foreman: string;
  peer: string;
  voiceSecretary?: string;
  actorNotes: Record<string, string>;
  actorOrder?: string[];
  extraTaggedBlocks?: string[];
}): string {
  const parts: string[] = [];
  const common = trimBlock(input.common);
  const foreman = trimBlock(input.foreman);
  const peer = trimBlock(input.peer);
  const voiceSecretary = trimBlock(input.voiceSecretary || "");
  const actorNotes = input.actorNotes || {};
  const extraTaggedBlocks = Array.isArray(input.extraTaggedBlocks) ? input.extraTaggedBlocks.map(trimBlock).filter(Boolean) : [];

  if (common) parts.push(common);
  if (foreman) parts.push(`## @role: foreman\n\n${foreman}`);
  if (peer) parts.push(`## @role: peer\n\n${peer}`);
  if (voiceSecretary) parts.push(`## @voice_secretary\n\n${voiceSecretary}`);

  const seen = new Set<string>();
  const orderedActorIds: string[] = [];
  for (const actorId of Array.isArray(input.actorOrder) ? input.actorOrder : []) {
    const id = String(actorId || "").trim();
    if (!id || seen.has(id)) continue;
    seen.add(id);
    orderedActorIds.push(id);
  }
  for (const actorId of Object.keys(actorNotes).sort()) {
    if (seen.has(actorId)) continue;
    seen.add(actorId);
    orderedActorIds.push(actorId);
  }
  for (const actorId of orderedActorIds) {
    const body = trimBlock(actorNotes[actorId]);
    if (!body) continue;
    parts.push(`## @actor: ${actorId}\n\n${body}`);
  }
  parts.push(...extraTaggedBlocks);
  const out = parts.filter(Boolean).join("\n\n").trim();
  return out ? `${out}\n` : "";
}

export function updateActorHelpNote(markdown: string, actorId: string, note: string, actorOrder?: string[]): string {
  const parsed = parseHelpMarkdown(markdown);
  const nextActorNotes = { ...parsed.actorNotes };
  const aid = String(actorId || "").trim();
  if (aid) nextActorNotes[aid] = trimBlock(note);
  if (aid && !nextActorNotes[aid]) delete nextActorNotes[aid];
  return buildHelpMarkdown({
    common: parsed.common,
    foreman: parsed.foreman,
    peer: parsed.peer,
    voiceSecretary: parsed.voiceSecretary,
    actorNotes: nextActorNotes,
    actorOrder,
    extraTaggedBlocks: parsed.extraTaggedBlocks,
  });
}

export function updateVoiceSecretaryHelpNote(markdown: string, note: string, actorOrder?: string[]): string {
  const parsed = parseHelpMarkdown(markdown);
  return buildHelpMarkdown({
    common: parsed.common,
    foreman: parsed.foreman,
    peer: parsed.peer,
    voiceSecretary: trimBlock(note),
    actorNotes: parsed.actorNotes,
    actorOrder,
    extraTaggedBlocks: parsed.extraTaggedBlocks,
  });
}
